rust/src/quic/frames.rs

```diff
@@ -115,13 +122,124 @@ impl fmt::Display for StreamTag {
+}
+
 #[derive(Debug, PartialEq)]
 pub(crate) enum Frame {
     Padding,
+    Ack(Ack),
+    Crypto(Crypto),
     Stream(Stream),
     Unknown(Vec<u8>),
 }
 
+fn parse_ack_frame(input: &[u8]) -> IResult<&[u8], Frame, QuicError> {
+    let rest = input;
+    let (rest, largest_acknowledged) = quic_var_uint(rest)?;
+    let (rest, ack_delay) = quic_var_uint(rest)?;
+    let (rest, ack_range_count) = quic_var_uint(rest)?;
+    let (rest, first_ack_range) = quic_var_uint(rest)?;
+
+    if ack_range_count != 0 {
+        //TODO RFC9000 section 19.3.1.  ACK Ranges
+        return Err(nom::Err::Error(QuicError::UnhandledYet));
+    }
+
+    Ok((
+        rest,
+        Frame::Ack(Ack {
+            largest_acknowledged,
+            ack_delay,
+            ack_range_count,
+            first_ack_range,
+        }),
+    ))
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct QuicTlsExtension {
+    pub etype: TlsExtensionType,
+    pub values: Vec<u8>,
+    pub indices: Vec<usize>,
+}
+
+// get interesting stuff out of parsed tls extensions
+fn quic_get_tls_extensions(input: Option<&[u8]>) -> Vec<QuicTlsExtension> {
+    let mut extv = Vec::new();
+    if let Some(extr) = input {
+        if let Ok((_, exts)) = parse_tls_extensions(extr) {
+            for e in &exts {
+                let etype = TlsExtensionType::from(e);
+                let mut values = Vec::new();
+                let mut indices = Vec::new();
+                match e {
+                    TlsExtension::SNI(x) => {
+                        for sni in x {
+                            values.extend_from_slice(sni.1);
+                            indices.push(values.len());
+                        }
+                    }
+                    TlsExtension::ALPN(x) => {
+                        for alpn in x {
+                            values.extend_from_slice(alpn);
+                            indices.push(values.len());
+                        }
+                    }
+                    _ => {}
+                }
+                extv.push(QuicTlsExtension {
+                    etype,
+                    values,
+                    indices,
+                })
+            }
+        }
+    }
+    return extv;
+}
+
+fn parse_crypto_frame(input: &[u8]) -> IResult<&[u8], Frame, QuicError> {
+    let rest = input;
+    let (rest, offset) = quic_var_uint(rest)?;
+    let (rest, length) = quic_var_uint(rest)?;
+    let (rest, _) = take(offset as usize)(rest)?;
+    let (_, data) = take(length as usize)(rest)?;
+
+    if let Ok((rest, msg)) = parse_tls_message_handshake(data) {
+        if let Handshake(hs) = msg {
+            match hs {
+                ClientHello(ch) => {
+                    let ciphers = ch.ciphers;
+                    let extv = quic_get_tls_extensions(ch.ext);
+                    return Ok((rest, Frame::Crypto(Crypto { ciphers, extv })));
+                }
+                ServerHello(sh) => {
+                    let ciphers = vec![sh.cipher];
```

Comment:
Do we want to log/detect the cipher ?
Wireshark gives a JA3 indeed by the way

Reply:
JA3 support would be nice, but I think we don't need to do it in a single step? I guess it depends on how much work it will be to support it.